File: src/main/java/info/szadkowski/contact/throttle/counter/TumblingCounter.java
package info.szadkowski.contact.throttle.counter;

import java.time.Duration;

public class TumblingCounter {
  private final Object lock = new Object();
  private final long windowSizeMillis;

  private long windowOpenTimeStampMillis = -1;
  private long counter = 0;

  public TumblingCounter(Duration windowSize) {
    this.windowSizeMillis = windowSize.toMillis();
  }

  public long count(long timeStampMillis) {
    synchronized (lock) {
      if (isNewWindow(timeStampMillis)) {
        windowOpenTimeStampMillis = timeStampMillis;
        counter = 0;
      }

      return ++counter;
    }
  }

  public boolean isNewWindow(long timeStampMillis) {
    synchronized (lock) {
      return isFirstWindow() || isTimeStampAfterWindow(timeStampMillis);
    }
  }

  private boolean isTimeStampAfterWindow(long timeStampMillis) {
    return windowOpenTimeStampMillis + windowSizeMillis <= timeStampMillis;
  }

  private boolean isFirstWindow() {
    return windowOpenTimeStampMillis < 0;
  }
}

File: src/main/kotlin/info/szadkowski/contact/throttle/counter/TumblingCounter.kt
package info.szadkowski.contact.throttle.counter

import java.time.Duration

class TumblingCounter(windowSize: Duration) {
    private val lock = Any()
    private val windowSizeMillis: Long = windowSize.toMillis()
    private var windowOpenTimeStampMillis: Long = -1
    private var counter: Long = 0

    private val isFirstWindow: Boolean
        get() = windowOpenTimeStampMillis < 0

    fun count(timeStampMillis: Long): Long {
        synchronized(lock) {
            if (isNewWindow(timeStampMillis)) {
                windowOpenTimeStampMillis = timeStampMillis
                counter = 0
            }

            return ++counter
        }
    }

    fun isNewWindow(timeStampMillis: Long): Boolean {
        synchronized(lock) {
            return isFirstWindow || isTimeStampAfterWindow(timeStampMillis)
        }
    }

    private fun isTimeStampAfterWindow(timeStampMillis: Long) =
        windowOpenTimeStampMillis + windowSizeMillis <= timeStampMillis
}
